### ISSUE TITLE
SPACK_DISABLE_LOCAL_CONFIG and bootstrapping in user space

### ISSUE THREAD

**Issue by tylerjereddy** [LLNL/spack#27704]
With `SPACK_DISABLE_LOCAL_CONFIG=1` I'm still seeing `spack` touching my user space when bootstrapping. For example, `spack` created `/users/treddy/.spack/bootstrap/config` and it was confusing me because it was not respecting my instance scope environment settings for compilers to use and so on.

Could we make `SPACK_DISABLE_LOCAL_CONFIG=1` an even stronger guarantee of ignoring/never writing to the local scope, even for bootstrapping?

**Comment by alalazo** [comment]
The default location used by Spack for bootstrapping can be changed with:
```console
$ spack bootstrap root <path>
```
and by default it is `~/.spack/bootstrap`. What we can do is, maybe, to move that into `$spack/opt` if `SPACK_DISABLE_LOCAL_CONFIG=1`. There are pros and cons in doing that (for instance different instances of Spack will need to bootstrap each on their own, rather than reusing binaries from `~/.spack/bootstrap`).

**Comment by haampie** [comment]
You should use `SPACK_USER_CACHE_PATH`, it's orthogonal to config

**Comment by tylerjereddy** [comment]
You're saying `SPACK_USER_CACHE_PATH` is controlling the writing to `/users/treddy/.spack/bootstrap/config` and despite `config` in the path name, that's actually a cache setting independent of the config?

**Comment by haampie** [comment]
Yeah, that does not sound right, but that's how it is right now. I think the config files you see there are supposed to be more of a log of what config was used during bootstrapping. I'm not sure if they are supposed to be used as bootstrapping configuration next time something is bootstrapped. @alalazo knows this better

**Comment by alalazo** [comment]
> I'm not sure if they are supposed to be used as bootstrapping configuration next time something is bootstrapped.

They should effectively act as a log (meaning I would consider any deviation from this behavior a bug). The idea is to minimize the amount of stuff that is directly configurable during bootstrapping, so that we can lower the amount of "edge-cases" we have to deal with.

What happens more in details during bootstrapping is:
1. A temporary configuration scope is created under `$bootstrap_root/config` (it will be the only one used during bootstrapping)
2. The `config` and `bootstrap` sections of the user configuration are copied over to the scope created at point 1.
3. Bootstrapping is performed for the software needed
4. In the end all the user's settings are restored

**Comment by alalazo** [comment]
To add on that, the function responsible for config substitution is:

https://github.com/spack/spack/blob/452a693aab98e6eaefe4a0ce078e4c2ef9c08e7e/lib/spack/spack/bootstrap.py#L672-L691
